authnServices.py

```diff
@@ -0,0 +1,317 @@
+        response.headers['Access-Control-Allow-Methods'] = "GET,OPTIONS"
+        response.headers['Access-Control-Allow-Headers'] = 'Origin, Accept, Authorization'
+        response["content_type"] = ct
+        response.status = status
+        return data
+
+    def not_implemented():
+        abort(501)
+
+    def dispatch_views(self):
+        # Handle getting user to log in
+        self.app.route('/{0}'.format(self.AUTH_URL_LOGIN), ["GET"], 
+            getattr(self.handler, "login", self.not_implemented))               
+
+        # Give back a cookie and close window
+        # Also where user comes back to after OAuth
+        self.app.route('/{0}'.format(self.AUTH_URL_COOKIE), ["GET"], 
+            getattr(self.handler, "cookie", self.not_implemented))
+
+        # PostMessage() or return a token
+        self.app.route('/{0}'.format(self.AUTH_URL_TOKEN), ["GET"], 
+            getattr(self.handler, "token", self.not_implemented))
+
+        # Let the user logout
+        if self.AUTH_URL_LOGOUT:
+            self.app.route('/{0}'.format(self.AUTH_URL_LOGOUT), ["GET"], 
+                getattr(self.handler, "logout", self.not_implemented)) 
+
+    def get_bottle_app(self):
+        """Returns bottle instance"""
+        self.app = Bottle()
+        self.dispatch_views()
+        return self.app    
+
+
+class AuthNHandler(object):
+
+    def __init__(self, app, cookie_secret, token_secret, salt, name="iiif_access_cookie",
+            cookie_domain="", cookie_path="/", is_https=True):
+        self.application = app
+        self.cookie_name = name
+        self.cookie_domain = cookie_domain
+        self.cookie_path = cookie_path
+        self.is_https = is_https
+        self.cookie_secret = cookie_secret
+        self.token_secret = token_secret
+        self.salt = salt
+
+    def kdf(self):
+       return PBKDF2HMAC(algorithm=hashes.SHA256(), length=32, salt=self.salt,
+            iterations=100000, backend=default_backend())
+
+    @staticmethod
+    def basic_origin(origin):
+        ccslds = ['co', 'org', 'com', 'net', 'ac', 'edu', 'gov', 'mil', 'or', \
+            'gen', 'govt', 'school', 'sch']
+        origin = origin.replace('https://', '')
+        origin = origin.replace('http://', '') # Mixed Content violation
+        slidx = origin.find("/")
+        if slidx > -1:
+            origin = origin[:slidx]
+        cidx = origin.find(":")
+        if cidx > -1:
+            origin = origin[:cidx]
+        domain = origin.split('.')
+        if len(domain) >= 3 and domain[-2] in ccslds:
+            # foo.x.cctld
+            secret = ".".join(domain[-3:])
+        elif domain[-1].isdigit():
+            # 10.0.0.1
+            secret = origin
+        elif len(domain) >= 2:
+            # foo.gtld
+            secret = ".".join(domain[-2:])
+        else:
+            # localhost or * ... hopefully not *!
+            secret = domain[0]
+        return secret
```

Comment:
This seems quite complicated and fragile – is there a reason you’re not just using `urlparse` from the standard library? At least for the test cases defined below, this code does the same thing:

```python
@staticmethod
def basic_origin(origin):
    u = urlparse(origin).hostname
    return u[len('www.'):] if u.startswith('www.') else u
```